firedrake/interpolation.py

```diff
@@ -124,31 +131,406 @@ class Interpolator(object):
-        """Compute the interpolation.
+    def interpolate(self, *args, **kwargs):
+        """
+        Compute the interpolation.
 
         :arg function: If the expression being interpolated contains an
             :class:`ufl.Argument`, then the :class:`.Function` value to
             interpolate.
         :kwarg output: Optional. A :class:`.Function` to contain the output.
         :kwarg transpose: Set to true to apply the transpose (adjoint) of the
               interpolation operator.
+        :kwarg default_missing_val: For interpolation across meshes: the
+            optional value to assign to DOFs in the target mesh that are
+            outside the source mesh. If this is not set and an ``output``
+            :class:`.Function` is specified, then such DOF values are left
+            unchanged. If this is not set and no ``output`` :class:`.Function`
+            is specified, then the default value is zero. This does not effect
+            transpose interpolation.
+
         :returns: The resulting interpolated :class:`.Function`.
         """
+        return self.interpolate(*args, **kwargs)
+
+
+class CrossMeshInterpolator(Interpolator):
+    """
+    Interpolate a function from one mesh and function space to another.
+
+    Only works when the function space we interpolate into has point evaluation
+    nodes. A more general version would require finding the global coordinates
+    of all quadrature points of the target function space in the source mesh.
+
+    For arguments, see :class:`.Interpolator`.
+    """
+
+    def __init__(
+        self,
+        expr,
+        V,
+        subset=None,
+        freeze_expr=False,
+        access=op2.WRITE,
+        bcs=None,
+        allow_missing_points=False,
+    ):
+        if subset:
+            raise NotImplementedError("subset not implemented")
+        if freeze_expr:
+            raise NotImplementedError("freeze_expr not implemented")
+        if access != op2.WRITE:
+            raise NotImplementedError("access other than op2.WRITE not implemented")
+        if bcs:
+            raise NotImplementedError("bcs not implemented")
+        if V.ufl_element().mapping() != "identity":
+            # Identity mapping between reference cell and physical coordinates
+            # implies point evaluation nodes
+            raise NotImplementedError(
+                "Can only interpolate into spaces with point evaluation nodes"
+            )
+
+        super().__init__(
+            expr, V, subset, freeze_expr, access, bcs, allow_missing_points
+        )
+
+        self.arguments = extract_arguments(expr)
+        self.nargs = len(self.arguments)
+        if self.nargs and self.arguments[0] != self.expr:
+            raise NotImplementedError(
+                "Can't yet create an interpolator from an expression with arguments."
+            )
+
+        if self._allow_missing_points:
+            missing_points_behaviour = None
```

Comment:
I think this would be better as an `enum`:

```py
class MissingPointsBehaviour(enum.Enum):
    NONE = enum.auto()  # not a great name
    ERROR = enum.auto()
```

Reply:
This is me specifying the necessary argument to `VertexOnlyMesh` (either `None` or `'error'`) which is part of the public API. I don't see how an `enum` is therefore helpful?

Reply:
I think generally if you have a finite number of options for something an enum is recommended. You can actually get away with this without changing the public API:

```py
class MissingPointsBehaviour(enum.Enum):
    NONE = None
    ERROR = "error"
```

then things like `missing_points_behaviour == MissingPointsBehaviour.ERROR` will work even if the user provides the string "error".

Reply:
Yes. The idea is to get rid of explicit strings in the internal code. You can also do some [straightforward validation of user input](https://stackoverflow.com/questions/43634618/how-do-i-test-if-int-value-exists-in-python-enum-without-using-try-catch).

I also think that `None`/`NONE` is not a very descriptive name. Should it be `"ignore"`/`IGNORE`?

Reply:
> I also think that `None`/`NONE` is not a very descriptive name. Should it be `"ignore"`/`IGNORE`?

Well it implies "No missing points behaviour", but if you don't think it's clear then that's fair enough. It's documented at least! But again I don't want to be changing the public API (and all the tests and manual pages) at this point.

How do enums interact with a public API in general? Is the user supposed to create an enum and supply that to the constructor? Does the displayed default value change to an instance of the enum? There are very few of these in the firedrake code base, none of which seem to be public facing, hence all my questions!

Reply:
Okay I've added the enum in dc4f2812245b2cb745d46b3a19c96b539cfc46ed